controllers/servingruntime_controller.go

```diff
@@ -281,11 +280,45 @@ func (r *ServingRuntimeReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{}, nil
 	}
 
-	replicas, requeueDuration, err := r.determineReplicasAndRequeueDuration(ctx, log, cfg, spec, req.NamespacedName)
+	var requeueDuration time.Duration
+	mmDeploymentName := fmt.Sprintf("%s-%s", mmDeployment.ServiceName, mmDeployment.Name)
+
+	var as *autoscaler.AutoscalerReconciler
+	if crt.GetName() != "" {
+		as, err = autoscaler.NewAutoscalerReconciler(r.Client, r.Client.Scheme(), crt, mmDeploymentName, mmDeployment.Namespace)
+	} else {
+		as, err = autoscaler.NewAutoscalerReconciler(r.Client, r.Client.Scheme(), rt, mmDeploymentName, mmDeployment.Namespace)
+	}
+
+	if err != nil {
+		log.Error(err, "fails to create an autoscaler controller: %w", "skip to create HPA")
+	}
+
+	replicas, requeueDurationRuntime, err := r.determineReplicasAndRequeueDuration(ctx, log, cfg, spec, req.NamespacedName)
 	if err != nil {
 		return RequeueResult, fmt.Errorf("could not determine replicas: %w", err)
 	}
+	requeueDuration = requeueDurationRuntime
 	mmDeployment.Replicas = replicas
+
+	//ScaleToZero or None autoscaler case
+	if replicas == uint16(0) || as.Autoscaler.AutoscalerClass == autoscaler.AutoscalerClassNone {
+		if _, err = as.Reconcile(true); err != nil {
+			return ctrl.Result{}, fmt.Errorf("HPA reconcile error: %w", err)
+		}
+	} else {
+		//Autoscaler case
+		if as.Autoscaler != nil {
+			//scaleUp
+			mmDeployment.Replicas = uint16(*(as.Autoscaler.HPA.HPA).Spec.MinReplicas)
```

Comment:
By setting this to the min replicas of the HPA and applying the Deployment, I think the Deployment will be scaled down to the minimum each time the ServingRuntime is reconciled. This can happen when the controller is restarted and will also happen periodically even without any change to the SR spec due to the [SyncPeriod](https://github.com/kubernetes-sigs/controller-runtime/blob/a26de2d610c3cf4b2a02688534aaf5a65749c743/pkg/manager/manager.go#L142-L169). If the HPA has scaled the Deployment up, the re-apply will cause a bunch of pods to terminate when replicas is set to the min and then new pods to spin up when the HPA corrects it (at least that's what I've seen in the past when manually applying replicas to a Deployment with an HPA).

That said, I'm not sure what the right solution is for this. Potentially, [server-side-apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) can help with its field level ownership, but we'd have to look into the details of how to implement it.
Another option may be to directly set the value for replicas outside of the `mmDeployment.Apply` and then skip setting the replicas when AutoScaling is active.

Reply:
You are right. The reason why I added this code is that HPA can not scale up the deployment object that has 0 replicas. Hence, I believe checking if the existing deployment's replicas are bigger than 0, we can skip the deployment.apply().

Reply:
Do you want me to refactor the mmDeployment with this PR? or would it be the next task? :) 
At the moment, I just add some tricks in there. If the deployment has more than 0 replicas, just set the same replicas  and apply the replicas. With the refactoring, this logic should be slightly changed.
please let me know your thought on this.

Reply:
Checking the replicas in the existing Deployment works too! I got the controller with your changes up and running and did some testing with it and it is working well (no expected changes to the replicas).

> Do you want me to refactor the mmDeployment with this PR? or would it be the next task? :)

Checking the replicas is working well, so we won't worry any follow up for now. Thanks!